README.md
## ct-adc-mini-msg

弱提示；兼容Vue插件的形式和普通js插件方式使用

## 组件示例图

![img](https://github.com/ct-adc/ct-adc-mini-msg/blob/dev/minimsg.png)

## 在线demo

[在线demo]()

## 使用

从npm安装ct-adc-mini-msg

```
npm install ct-adc-mini-msg --save
```
在代码中使用

```
import Vue from 'vue';
import MiniMsg from 'ct-adc-mini-msg';

Vue.use(MiniMsg);
在组件中使用:
this.$minimsg(options);

或者

new MiniMsg(options).animation();
```

## options

参数 | 说明 | 类型 | 默认值 | 可选值 | 描述 |
--- | --- | --- | --- | ---- | ---
content | 提示内容 | String | '' | 字符串（支持html渲染）|
type | 提示类型 | String | 'info' | success/error/warning/info | 
container | 承载信息框的外层容器 | DOM | body | DOM元素 |
stay | 信息框的停留时长，单位:秒 | Number | 1 | 数字 |
duration | 显示/隐藏动画的时长，单位:秒 | Number | 0.3 | 数字 |
top | 信息框容器顶部的最大高度 | Number | 16 | 数字 |

## this.$minimsg

弹出一个弱提示

### 参数

同options

### 返回值

MiniMsg实例


## 方法 （MiniMsg实例的方法）

### animation

执行信息框动画，通过执行该方法来显示一个信息框。

#### 参数列表

参数 | 说明 | 类型 | 默认值 | 可选值 | 描述 |
--- | --- | --- | --- | ---- | ----
callback 非必填 执行完动画后的回调函数

### 返回值

类型: this

### destroy

实时销毁minimsg

#### 参数列表

无参数

### 返回值

undefined

## 更新日志

[更新日志](https://github.com/ct-adc/ct-adc-mini-msg/blob/dev/CHANGELOG.md)

## 外部资源依赖列表

jquery



CHANGELOG.md
## 更新日志

### 2.0.0

*2018-07-24*

因ct-adc-mini-msg之前有相同名称的组件，故该组件需升级至2.0.0

### 1.0.0-alpha.3

*2018-06-05*

- 新增 添加stay参数 

- 修改 修改duration为stay

### 1.0.0-alpha.2

**2018-05-28**

- 修复 样式没有被正确使用

### 1.0.0-alpha.1

**2018-04-27**

发布1.0.0-alpha.1

src/component/main.js
import './mini-msg';
const MiniMsg = function(ops) {
    const defaults = {
        content: '',
        type: 'info',
        container: document.body,
        stay: 1,
        duration: 0.2,
        top: 16
    };

    $.extend(this, defaults, ops);
    this.init();
};

MiniMsg.prototype = {
    constructor: MiniMsg,
    classForType: {
        success: 'mini-msg-success',
        warning: 'mini-msg-warning',
        error: 'mini-msg-error',
        info: 'mini-msg-info'
    },
    iconForType: {
        success: 'glyphicon-ok-sign',
        warning: 'glyphicon-exclamation-sign',
        error: 'glyphicon-remove-circle',
        info: 'glyphicon-info-sign'
    },
    /**
     * 设置msgBox的样式来显示/隐藏它
     * @param {string} showOrHide "show" | "hide"
     */
    setBoxStyle: function(showOrHide) {
        const style = {};

        if (showOrHide === 'show') {
            style.transform = `translateY(${this.top}px)`;
            style.opacity = 1;
        } else {
            style.transform = `translateY(-100%)`;
            style.opacity = 0;
        }
        this.msgBox.css(style);
    },
    init: function() {
        //  生成一个dom，并插入到页面中
        //  根据type设置它的样式
        const html = `<div class="mini-msg" style="transition: transform ${this.duration}s, opacity ${this.duration}s;">
             <div class="mini-msg-notice">
            <div class="mini-msg-notice-content ${this.classForType[this.type]}">
            <i class="glyphicon ${this.iconForType[this.type]}"></i>
            ${this.content}
            </div>
            </div>
            </div>`;
        const $miniMsg = $(html);
        let styleForContainer;

        $(this.container).append($miniMsg);

        if (this.container !== document.body) {
            styleForContainer = {
                'position': 'relative',
                'overflow-x': 'hidden'
            };
        } else {
            styleForContainer = {
                'overflow-x': 'hidden'
            };
        }
        $(this.container).css(styleForContainer);
        this.msgBox = $miniMsg;
        this.setBoxStyle('hide');
    },
    show: function() {
        //显示
        this.setBoxStyle('show');
    },
    hide: function() {
        //隐藏
        this.setBoxStyle('hide');
    },
    destroy: function() {
        //销毁
        if (typeof this.timer !== 'undefined') {
            clearTimeout(this.timer);
        }
        this.msgBox.remove();
    },
    /**
     * 执行动画
     * @param {Function} callback 执行完之后的回调函数
     */
    animation: function(callback) {
        const that = this;

        that.timer = setTimeout(()=>{
            this.show();
            this.timer = setTimeout(()=>{
                this.hide();
                this.timer = setTimeout(()=>{
                    if (typeof callback === 'function') {
                        callback();
                    }
                    this.destroy();
                }, 1000 * this.duration);
            }, 1000 * this.duration + 1000 * this.stay);
        }, 1000 * this.duration);
        return this;
    }
};

MiniMsg.install = function(Vue) {
    Vue.prototype.$minimsg = function(options, callback) {
        return new MiniMsg(options).animation(() => {
            if (typeof callback === 'function') {
                callback();
            }
        });
    };
};
export default MiniMsg;


lib/index.js
(function webpackUniversalModuleDefinition(root, factory) {
	if(typeof exports === 'object' && typeof module === 'object')
		module.exports = factory();
	else if(typeof define === 'function' && define.amd)
		define([], factory);
	else if(typeof exports === 'object')
		exports["ct-adc-mini-msg"] = factory();
	else
		root["ct-adc-mini-msg"] = factory();
})(typeof self !== 'undefined' ? self : this, function() {
return /******/ (function(modules) { // webpackBootstrap
/******/ 	// The module cache
/******/ 	var installedModules = {};
/******/
/******/ 	// The require function
/******/ 	function __webpack_require__(moduleId) {
/******/
/******/ 		// Check if module is in cache
/******/ 		if(installedModules[moduleId]) {
/******/ 			return installedModules[moduleId].exports;
/******/ 		}
/******/ 		// Create a new module (and put it into the cache)
/******/ 		var module = installedModules[moduleId] = {
/******/ 			i: moduleId,
/******/ 			l: false,
/******/ 			exports: {}
/******/ 		};
/******/
/******/ 		// Execute the module function
/******/ 		modules[moduleId].call(module.exports, module, module.exports, __webpack_require__);
/******/
/******/ 		// Flag the module as loaded
/******/ 		module.l = true;
/******/
/******/ 		// Return the exports of the module
/******/ 		return module.exports;
/******/ 	}
/******/
/******/
/******/ 	// expose the modules object (__webpack_modules__)
/******/ 	__webpack_require__.m = modules;
/******/
/******/ 	// expose the module cache
/******/ 	__webpack_require__.c = installedModules;
/******/
/******/ 	// define getter function for harmony exports
/******/ 	__webpack_require__.d = function(exports, name, getter) {
/******/ 		if(!__webpack_require__.o(exports, name)) {
/******/ 			Object.defineProperty(exports, name, {
/******/ 				configurable: false,
/******/ 				enumerable: true,
/******/ 				get: getter
/******/ 			});
/******/ 		}
/******/ 	};
/******/
/******/ 	// getDefaultExport function for compatibility with non-harmony modules
/******/ 	__webpack_require__.n = function(module) {
/******/ 		var getter = module && module.__esModule ?
/******/ 			function getDefault() { return module['default']; } :
/******/ 			function getModuleExports() { return module; };
/******/ 		__webpack_require__.d(getter, 'a', getter);
/******/ 		return getter;
/******/ 	};
/******/
/******/ 	// Object.prototype.hasOwnProperty.call
/******/ 	__webpack_require__.o = function(object, property) { return Object.prototype.hasOwnProperty.call(object, property); };
/******/
/******/ 	// __webpack_public_path__
/******/ 	__webpack_require__.p = "";
/******/
/******/ 	// Load entry module and return exports
/******/ 	return __webpack_require__(__webpack_require__.s = 0);
/******/ })
/************************************************************************/
/******/ ([
/* 0 */
/***/ (function(module, exports, __webpack_require__) {

"use strict";


Object.defineProperty(exports, "__esModule", {
    value: true
});

__webpack_require__(1);

var MiniMsg = function MiniMsg(ops) {
    var defaults = {
        content: '',
        type: 'info',
        container: document.body,
        stay: 1,
        duration: 0.2,
        top: 16
    };

    $.extend(this, defaults, ops);
    this.init();
};

MiniMsg.prototype = {
    constructor: MiniMsg,
    classForType: {
        success: 'mini-msg-success',
        warning: 'mini-msg-warning',
        error: 'mini-msg-error',
        info: 'mini-msg-info'
    },
    iconForType: {
        success: 'glyphicon-ok-sign',
        warning: 'glyphicon-exclamation-sign',
        error: 'glyphicon-remove-circle',
        info: 'glyphicon-info-sign'
    },

    setBoxStyle: function setBoxStyle(showOrHide) {
        var style = {};

        if (showOrHide === 'show') {
            style.transform = 'translateY(' + this.top + 'px)';
            style.opacity = 1;
        } else {
            style.transform = 'translateY(-100%)';
            style.opacity = 0;
        }
        this.msgBox.css(style);
    },
    init: function init() {
        var html = '<div class="mini-msg" style="transition: transform ' + this.duration + 's, opacity ' + this.duration + 's;">\n             <div class="mini-msg-notice">\n            <div class="mini-msg-notice-content ' + this.classForType[this.type] + '">\n            <i class="glyphicon ' + this.iconForType[this.type] + '"></i>\n            ' + this.content + '\n            </div>\n            </div>\n            </div>';
        var $miniMsg = $(html);
        var styleForContainer = void 0;

        $(this.container).append($miniMsg);

        if (this.container !== document.body) {
            styleForContainer = {
                'position': 'relative',
                'overflow-x': 'hidden'
            };
        } else {
            styleForContainer = {
                'overflow-x': 'hidden'
            };
        }
        $(this.container).css(styleForContainer);
        this.msgBox = $miniMsg;
        this.setBoxStyle('hide');
    },
    show: function show() {
        this.setBoxStyle('show');
    },
    hide: function hide() {
        this.setBoxStyle('hide');
    },
    destroy: function destroy() {
        if (typeof this.timer !== 'undefined') {
            clearTimeout(this.timer);
        }
        this.msgBox.remove();
    },

    animation: function animation(callback) {
        var _this = this;

        var that = this;

        that.timer = setTimeout(function () {
            _this.show();
            _this.timer = setTimeout(function () {
                _this.hide();
                _this.timer = setTimeout(function () {
                    if (typeof callback === 'function') {
                        callback();
                    }
                    _this.destroy();
                }, 1000 * _this.duration);
            }, 1000 * _this.duration + 1000 * _this.stay);
        }, 1000 * this.duration);
        return this;
    }
};

MiniMsg.install = function (Vue) {
    Vue.prototype.$minimsg = function (options, callback) {
        return new MiniMsg(options).animation(function () {
            if (typeof callback === 'function') {
                callback();
            }
        });
    };
};
exports.default = MiniMsg;

/***/ }),
/* 1 */
/***/ (function(module, exports, __webpack_require__) {


var content = __webpack_require__(2);

if(typeof content === 'string') content = [[module.i, content, '']];

var transform;
var insertInto;



var options = {"hmr":true}

options.transform = transform
options.insertInto = undefined;

var update = __webpack_require__(4)(content, options);

if(content.locals) module.exports = content.locals;

if(false) {
	module.hot.accept("!!../../node_modules/css-loader/index.js!./mini-msg.css", function() {
		var newContent = require("!!../../node_modules/css-loader/index.js!./mini-msg.css");

		if(typeof newContent === 'string') newContent = [[module.id, newContent, '']];

		var locals = (function(a, b) {
			var key, idx = 0;

			for(key in a) {
				if(!b || a[key] !== b[key]) return false;
				idx++;
			}

			for(key in b) idx--;

			return idx === 0;
		}(content.locals, newContent.locals));

		if(!locals) throw new Error('Aborting CSS HMR due to changed css-modules locals.');

		update(newContent);
	});

	module.hot.dispose(function() { update(); });
}

/***/ }),
/* 2 */
/***/ (function(module, exports, __webpack_require__) {

exports = module.exports = __webpack_require__(3)();
// imports


// module
exports.push([module.i, ".mini-msg {\n    border-radius: 5px;\n    font-size: 12px;\n    position: fixed;\n    z-index: 9999;\n    width: 100%;\n    top: 0;\n    left: 0;\n    height: auto;\n    background-color: #ccc;\n    color:#fff;\n    opacity: 0;\n    transform: translateY(-100%);\n}\n\n.mini-msg-notice {\n    width: auto;\n    vertical-align: middle;\n    position: absolute;\n    left: 50%;\n}\n\n.mini-msg-notice-content {\n    position: relative;\n    right: 50%;\n    padding: 8px 16px 8px 32px;\n    border-radius: 6px;\n    box-shadow: 0 2px 8px rgba(0, 0, 0, 0.2);\n    background: #fff;\n    display: block;\n}\n\n.mini-msg-notice-content .glyphicon {\n    margin-right: 8px;\n    font-size: 14px;\n    position: absolute;\n    top: 50%;\n    left: 14px;\n    -webkit-transform: translateY(-50%);\n    -moz-transform: translateY(-50%);\n    -ms-transform: translateY(-50%);\n    -o-transform: translateY(-50%);\n    transform: translateY(-50%);\n}\n\n.mini-msg-success {\n    background-color:#5cb85c;\n}\n\n.mini-msg-warning {\n    background-color:#f0ad4e;\n}\n\n.mini-msg-error {\n    background-color:#d9534f;\n}\n\n.mini-msg-info {\n    background-color:#5bc0de;\n}\n", ""]);

// exports


/***/ }),
/* 3 */
/***/ (function(module, exports) {

/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/
// css base code, injected by the css-loader
module.exports = function() {
	var list = [];

	// return the list of modules as css string
	list.toString = function toString() {
		var result = [];
		for(var i = 0; i < this.length; i++) {
			var item = this[i];
			if(item[2]) {
				result.push("@media " + item[2] + "{" + item[1] + "}");
			} else {
				result.push(item[1]);
			}
		}
		return result.join("");
	};

	// import a list of modules into the list
	list.i = function(modules, mediaQuery) {
		if(typeof modules === "string")
			modules = [[null, modules, ""]];
		var alreadyImportedModules = {};
		for(var i = 0; i < this.length; i++) {
			var id = this[i][0];
			if(typeof id === "number")
				alreadyImportedModules[id] = true;
		}
		for(i = 0; i < modules.length; i++) {
			var item = modules[i];
			// skip already imported module
			// this implementation is not 100% perfect for weird media query combinations
			//  when a module is imported multiple times with different media queries.
			//  I hope this will never occur (Hey this way we have smaller bundles)
			if(typeof item[0] !== "number" || !alreadyImportedModules[item[0]]) {
				if(mediaQuery && !item[2]) {
					item[2] = mediaQuery;
				} else if(mediaQuery) {
					item[2] = "(" + item[2] + ") and (" + mediaQuery + ")";
				}
				list.push(item);
			}
		}
	};
	return list;
};


/***/ }),
/* 4 */
/***/ (function(module, exports, __webpack_require__) {

/*
	MIT License http://www.opensource.org/licenses/mit-license.php
	Author Tobias Koppers @sokra
*/

var stylesInDom = {};

var	memoize = function (fn) {
	var memo;

	return function () {
		if (typeof memo === "undefined") memo = fn.apply(this, arguments);
		return memo;
	};
};

var isOldIE = memoize(function () {
	// Test for IE <= 9 as proposed by Browserhacks
	// @see http://browserhacks.com/#hack-e71d8692f65334173fee715c222cb805
	// Tests for existence of standard globals is to allow style-loader
	// to operate correctly into non-standard environments
	// @see https://github.com/webpack-contrib/style-loader/issues/177
	return window && document && document.all && !window.atob;
});

var getTarget = function (target) {
  return document.querySelector(target);
};

var getElement = (function (fn) {
	var memo = {};

	return function(target) {
                // If passing function in options, then use it for resolve "head" element.
                // Useful for Shadow Root style i.e
                // {
                //   insertInto: function () { return document.querySelector("#foo").shadowRoot }
                // }
                if (typeof target === 'function') {
                        return target();
                }
                if (typeof memo[target] === "undefined") {
			var styleTarget = getTarget.call(this, target);
			// Special case to return head of iframe instead of iframe itself
			if (window.HTMLIFrameElement && styleTarget instanceof window.HTMLIFrameElement) {
				try {
					// This will throw an exception if access to iframe is blocked
					// due to cross-origin restrictions
					styleTarget = styleTarget.contentDocument.head;
				} catch(e) {
					styleTarget = null;
				}
			}
			memo[target] = styleTarget;
		}
		return memo[target]
	};
})();

var singleton = null;
var	singletonCounter = 0;
var	stylesInsertedAtTop = [];

var	fixUrls = __webpack_require__(5);

module.exports = function(list, options) {
	if (typeof DEBUG !== "undefined" && DEBUG) {
		if (typeof document !== "object") throw new Error("The style-loader cannot be used in a non-browser environment");
	}

	options = options || {};

	options.attrs = typeof options.attrs === "object" ? options.attrs : {};

	// Force single-tag solution on IE6-9, which has a hard limit on the # of <style>
	// tags it will allow on a page
	if (!options.singleton && typeof options.singleton !== "boolean") options.singleton = isOldIE();

	// By default, add <style> tags to the <head> element
        if (!options.insertInto) options.insertInto = "head";

	// By default, add <style> tags to the bottom of the target
	if (!options.insertAt) options.insertAt = "bottom";

	var styles = listToStyles(list, options);

	addStylesToDom(styles, options);

	return function update (newList) {
		var mayRemove = [];

		for (var i = 0; i < styles.length; i++) {
			var item = styles[i];
			var domStyle = stylesInDom[item.id];

			domStyle.refs--;
			mayRemove.push(domStyle);
		}

		if(newList) {
			var newStyles = listToStyles(newList, options);
			addStylesToDom(newStyles, options);
		}

		for (var i = 0; i < mayRemove.length; i++) {
			var domStyle = mayRemove[i];

			if(domStyle.refs === 0) {
				for (var j = 0; j < domStyle.parts.length; j++) domStyle.parts[j]();

				delete stylesInDom[domStyle.id];
			}
		}
	};
};

function addStylesToDom (styles, options) {
	for (var i = 0; i < styles.length; i++) {
		var item = styles[i];
		var domStyle = stylesInDom[item.id];

		if(domStyle) {
			domStyle.refs++;

			for(var j = 0; j < domStyle.parts.length; j++) {
				domStyle.parts[j](item.parts[j]);
			}

			for(; j < item.parts.length; j++) {
				domStyle.parts.push(addStyle(item.parts[j], options));
			}
		} else {
			var parts = [];

			for(var j = 0; j < item.parts.length; j++) {
				parts.push(addStyle(item.parts[j], options));
			}

			stylesInDom[item.id] = {id: item.id, refs: 1, parts: parts};
		}
	}
}

function listToStyles (list, options) {
	var styles = [];
	var newStyles = {};

	for (var i = 0; i < list.length; i++) {
		var item = list[i];
		var id = options.base ? item[0] + options.base : item[0];
		var css = item[1];
		var media = item[2];
		var sourceMap = item[3];
		var part = {css: css, media: media, sourceMap: sourceMap};

		if(!newStyles[id]) styles.push(newStyles[id] = {id: id, parts: [part]});
		else newStyles[id].parts.push(part);
	}

	return styles;
}

function insertStyleElement (options, style) {
	var target = getElement(options.insertInto)

	if (!target) {
		throw new Error("Couldn't find a style target. This probably means that the value for the 'insertInto' parameter is invalid.");
	}

	var lastStyleElementInsertedAtTop = stylesInsertedAtTop[stylesInsertedAtTop.length - 1];

	if (options.insertAt === "top") {
		if (!lastStyleElementInsertedAtTop) {
			target.insertBefore(style, target.firstChild);
		} else if (lastStyleElementInsertedAtTop.nextSibling) {
			target.insertBefore(style, lastStyleElementInsertedAtTop.nextSibling);
		} else {
			target.appendChild(style);
		}
		stylesInsertedAtTop.push(style);
	} else if (options.insertAt === "bottom") {
		target.appendChild(style);
	} else if (typeof options.insertAt === "object" && options.insertAt.before) {
		var nextSibling = getElement(options.insertInto + " " + options.insertAt.before);
		target.insertBefore(style, nextSibling);
	} else {
		throw new Error("[Style Loader]\n\n Invalid value for parameter 'insertAt' ('options.insertAt') found.\n Must be 'top', 'bottom', or Object.\n (https://github.com/webpack-contrib/style-loader#insertat)\n");
	}
}

function removeStyleElement (style) {
	if (style.parentNode === null) return false;
	style.parentNode.removeChild(style);

	var idx = stylesInsertedAtTop.indexOf(style);
	if(idx >= 0) {
		stylesInsertedAtTop.splice(idx, 1);
	}
}

function createStyleElement (options) {
	var style = document.createElement("style");

	if(options.attrs.type === undefined) {
		options.attrs.type = "text/css";
	}

	addAttrs(style, options.attrs);
	insertStyleElement(options, style);

	return style;
}

function createLinkElement (options) {
	var link = document.createElement("link");

	if(options.attrs.type === undefined) {
		options.attrs.type = "text/css";
	}
	options.attrs.rel = "stylesheet";

	addAttrs(link, options.attrs);
	insertStyleElement(options, link);

	return link;
}

function addAttrs (el, attrs) {
	Object.keys(attrs).forEach(function (key) {
		el.setAttribute(key, attrs[key]);
	});
}

function addStyle (obj, options) {
	var style, update, remove, result;

	// If a transform function was defined, run it on the css
	if (options.transform && obj.css) {
	    result = options.transform(obj.css);

	    if (result) {
	    	// If transform returns a value, use that instead of the original css.
	    	// This allows running runtime transformations on the css.
	    	obj.css = result;
	    } else {
	    	// If the transform function returns a falsy value, don't add this css.
	    	// This allows conditional loading of css
	    	return function() {
	    		// noop
	    	};
	    }
	}

	if (options.singleton) {
		var styleIndex = singletonCounter++;

		style = singleton || (singleton = createStyleElement(options));

		update = applyToSingletonTag.bind(null, style, styleIndex, false);
		remove = applyToSingletonTag.bind(null, style, styleIndex, true);

	} else if (
		obj.sourceMap &&
		typeof URL === "function" &&
		typeof URL.createObjectURL === "function" &&
		typeof URL.revokeObjectURL === "function" &&
		typeof Blob === "function" &&
		typeof btoa === "function"
	) {
		style = createLinkElement(options);
		update = updateLink.bind(null, style, options);
		remove = function () {
			removeStyleElement(style);

			if(style.href) URL.revokeObjectURL(style.href);
		};
	} else {
		style = createStyleElement(options);
		update = applyToTag.bind(null, style);
		remove = function () {
			removeStyleElement(style);
		};
	}

	update(obj);

	return function updateStyle (newObj) {
		if (newObj) {
			if (
				newObj.css === obj.css &&
				newObj.media === obj.media &&
				newObj.sourceMap === obj.sourceMap
			) {
				return;
			}

			update(obj = newObj);
		} else {
			remove();
		}
	};
}

var replaceText = (function () {
	var textStore = [];

	return function (index, replacement) {
		textStore[index] = replacement;

		return textStore.filter(Boolean).join('\n');
	};
})();

function applyToSingletonTag (style, index, remove, obj) {
	var css = remove ? "" : obj.css;

	if (style.styleSheet) {
		style.styleSheet.cssText = replaceText(index, css);
	} else {
		var cssNode = document.createTextNode(css);
		var childNodes = style.childNodes;

		if (childNodes[index]) style.removeChild(childNodes[index]);

		if (childNodes.length) {
			style.insertBefore(cssNode, childNodes[index]);
		} else {
			style.appendChild(cssNode);
		}
	}
}

function applyToTag (style, obj) {
	var css = obj.css;
	var media = obj.media;

	if(media) {
		style.setAttribute("media", media)
	}

	if(style.styleSheet) {
		style.styleSheet.cssText = css;
	} else {
		while(style.firstChild) {
			style.removeChild(style.firstChild);
		}

		style.appendChild(document.createTextNode(css));
	}
}

function updateLink (link, options, obj) {
	var css = obj.css;
	var sourceMap = obj.sourceMap;

	/*
		If convertToAbsoluteUrls isn't defined, but sourcemaps are enabled
		and there is no publicPath defined then lets turn convertToAbsoluteUrls
		on by default.  Otherwise default to the convertToAbsoluteUrls option
		directly
	*/
	var autoFixUrls = options.convertToAbsoluteUrls === undefined && sourceMap;

	if (options.convertToAbsoluteUrls || autoFixUrls) {
		css = fixUrls(css);
	}

	if (sourceMap) {
		// http://stackoverflow.com/a/26603875
		css += "\n/*# sourceMappingURL=data:application/json;base64," + btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap)))) + " */";
	}

	var blob = new Blob([css], { type: "text/css" });

	var oldSrc = link.href;

	link.href = URL.createObjectURL(blob);

	if(oldSrc) URL.revokeObjectURL(oldSrc);
}


/***/ }),
/* 5 */
/***/ (function(module, exports) {


/**
 * When source maps are enabled, `style-loader` uses a link element with a data-uri to
 * embed the css on the page. This breaks all relative urls because now they are relative to a
 * bundle instead of the current page.
 *
 * One solution is to only use full urls, but that may be impossible.
 *
 * Instead, this function "fixes" the relative urls to be absolute according to the current page location.
 *
 * A rudimentary test suite is located at `test/fixUrls.js` and can be run via the `npm test` command.
 *
 */

module.exports = function (css) {
  // get current location
  var location = typeof window !== "undefined" && window.location;

  if (!location) {
    throw new Error("fixUrls requires window.location");
  }

	// blank or null?
	if (!css || typeof css !== "string") {
	  return css;
  }

  var baseUrl = location.protocol + "//" + location.host;
  var currentDir = baseUrl + location.pathname.replace(/\/[^\/]*$/, "/");

	// convert each url(...)
	/*
	This regular expression is just a way to recursively match brackets within
	a string.

	 /url\s*\(  = Match on the word "url" with any whitespace after it and then a parens
	   (  = Start a capturing group
	     (?:  = Start a non-capturing group
	         [^)(]  = Match anything that isn't a parentheses
	         |  = OR
	         \(  = Match a start parentheses
	             (?:  = Start another non-capturing groups
	                 [^)(]+  = Match anything that isn't a parentheses
	                 |  = OR
	                 \(  = Match a start parentheses
	                     [^)(]*  = Match anything that isn't a parentheses
	                 \)  = Match a end parentheses
	             )  = End Group
              *\) = Match anything and then a close parens
          )  = Close non-capturing group
          *  = Match anything
       )  = Close capturing group
	 \)  = Match a close parens

	 /gi  = Get all matches, not the first.  Be case insensitive.
	 */
	var fixedCss = css.replace(/url\s*\(((?:[^)(]|\((?:[^)(]+|\([^)(]*\))*\))*)\)/gi, function(fullMatch, origUrl) {
		// strip quotes (if they exist)
		var unquotedOrigUrl = origUrl
			.trim()
			.replace(/^"(.*)"$/, function(o, $1){ return $1; })
			.replace(/^'(.*)'$/, function(o, $1){ return $1; });

		// already a full url? no change
		if (/^(#|data:|http:\/\/|https:\/\/|file:\/\/\/|\s*$)/i.test(unquotedOrigUrl)) {
		  return fullMatch;
		}

		// convert the url to a full url
		var newUrl;

		if (unquotedOrigUrl.indexOf("//") === 0) {
		  	//TODO: should we add protocol?
			newUrl = unquotedOrigUrl;
		} else if (unquotedOrigUrl.indexOf("/") === 0) {
			// path should be relative to the base url
			newUrl = baseUrl + unquotedOrigUrl; // already starts with '/'
		} else {
			// path should be relative to current directory
			newUrl = currentDir + unquotedOrigUrl.replace(/^\.\//, ""); // Strip leading './'
		}

		// send back the fixed url(...)
		return "url(" + JSON.stringify(newUrl) + ")";
	});

	// send back the fixed css
	return fixedCss;
};


/***/ })
/******/ ]);
});